### PR TITLE
Surface send_hello server errors

### PR DESCRIPTION
**Pull request by urianchang**
I was having an issue catching OPC UA server error responses when trying to create an OPC UA client because the `UaStatusCodeError` would be logged in the socket thread and a `concurrent.futures.TimeoutError` would get thrown in the main process. By re-raising the `UaStatusCodeError`, the main process can catch it and provide more details as to why the client connection can not be established.

## Environment
**Server:** Instance of Prosys OPC UA Simulation Server in a Docker container. Set the Server Name to `pika`.

**Client code:**
```python
# Python 3.6
import concurrent.futures
import logging

import opcua
from opcua.ua import status_codes
from opcua.ua.uaerrors import UaStatusCodeError


logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s.%(msecs)03d [%(name)s] %(thread)d %(levelname)s - %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S"
)
logging.getLogger("opcua").setLevel(logging.DEBUG)
log = logging.getLogger(__name__)


host = "localhost"
server_name = "missingno"   # should throw an error
url = f"opc.tcp://{host}:4906/{server_name}"


try:
    with opcua.Client(url, timeout=1) as client:
        log.info("Connection successful!")
except concurrent.futures.TimeoutError:
    log.info("Hmm... Socket timed out")
except UaStatusCodeError as e:
    name, doc = status_codes.get_name_and_doc(e.code)
    log.info(f"OPC UA server responded with error {name}: {doc}")
``` 

Logs from current behavior:
```
cryptography is not installed, use of crypto disabled
cryptography is not installed, use of crypto disabled
cryptography is not installed, use of crypto disabled
2018-12-04 17:18:51.071 [opcua.client.ua_client.Socket] 140735838761856 INFO - opening connection
2018-12-04 17:18:51.075 [opcua.client.ua_client.Socket] 123145344372736 INFO - Thread started
2018-12-04 17:18:51.075 [opcua.uaprotocol] 123145344372736 DEBUG - Waiting for header
2018-12-04 17:18:51.081 [opcua.uaprotocol] 123145344372736 INFO - received header: Header(type:b'ERR', chunk_type:b'F', body_size:123, channel:0)
2018-12-04 17:18:51.082 [opcua.uaprotocol] 123145344372736 WARNING - Received an error: MessageAbort(error:StatusCode(Good), reason:Bad_TcpEndpointUrlInvalid (code=0x80830000, description="The Server does not recognize the QueryString specified."))
2018-12-04 17:18:51.082 [opcua.client.ua_client.Socket] 123145344372736 WARNING - Received an error: MessageAbort(error:StatusCode(Good), reason:Bad_TcpEndpointUrlInvalid (code=0x80830000, description="The Server does not recognize the QueryString specified."))
2018-12-04 17:18:51.082 [opcua.uaprotocol] 123145344372736 DEBUG - Waiting for header
2018-12-04 17:18:51.082 [opcua.client.ua_client.Socket] 123145344372736 INFO - Socket has closed connection
2018-12-04 17:18:51.082 [opcua.client.ua_client.Socket] 123145344372736 INFO - Thread ended
2018-12-04 17:18:52.079 [opcua.client.ua_client.Socket] 140735838761856 INFO - Request to close socket received
2018-12-04 17:18:52.080 [opcua.client.ua_client.Socket] 140735838761856 INFO - Socket closed, waiting for receiver thread to terminate...
2018-12-04 17:18:52.080 [opcua.client.ua_client.Socket] 140735838761856 INFO - Done closing socket: Receiving thread terminated, socket disconnected
2018-12-04 17:18:52.080 [__main__] 140735838761856 INFO - Hmm... Socket timed out
```

Logs with changes:
```
cryptography is not installed, use of crypto disabled
cryptography is not installed, use of crypto disabled
cryptography is not installed, use of crypto disabled
2018-12-04 17:20:00.544 [opcua.client.ua_client.Socket] 140735838761856 INFO - opening connection
2018-12-04 17:20:00.548 [opcua.client.ua_client.Socket] 123145328926720 INFO - Thread started
2018-12-04 17:20:00.548 [opcua.uaprotocol] 123145328926720 DEBUG - Waiting for header
2018-12-04 17:20:00.556 [opcua.uaprotocol] 123145328926720 INFO - received header: Header(type:b'ERR', chunk_type:b'F', body_size:123, channel:0)
2018-12-04 17:20:00.556 [opcua.uaprotocol] 123145328926720 DEBUG - Waiting for header
2018-12-04 17:20:00.556 [__main__] 140735838761856 INFO - OPC UA server responded with error BadTcpEndpointUrlInvalid: The server does not recognize the QueryString specified.
2018-12-04 17:20:00.557 [opcua.client.ua_client.Socket] 123145328926720 INFO - Socket has closed connection
2018-12-04 17:20:00.557 [opcua.client.ua_client.Socket] 123145328926720 INFO - Thread ended
```

### CODE REVIEW

File: opcua/client/client.py [hunk 1]
```diff
@@ -256,6 +267,10 @@ def connect(self):
         self.connect_socket()
         try:
             self.send_hello()
+        except ua.UaStatusCodeError:
+            raise
+
+        try:
```
Comment (oroulet) [inline opcua/client/client.py]: by moving send_hello in its own try/Catch you are not closing correctly the socket

Comment (oroulet) [reply opcua/client/client.py]: and catching an exception for then re-raising it immediatly without doing anything, does not make sense

File: opcua/client/client.py [hunk 2]
```diff
@@ -288,7 +303,10 @@ def send_hello(self):
         Send OPC-UA hello to server
         """
         ack = self.uaclient.send_hello(self.server_url.geturl(), self.max_messagesize, self.max_chunkcount)
-        # FIXME check ack
+
+        # TODO: Handle ua.UaError
+        if isinstance(ack, ua.UaStatusCodeError):
+            raise ack
```
Comment (oroulet) [inline opcua/client/client.py]: this might be correct, I am not sure, I do not remember what send_hello returns...

File: opcua/client/ua_client.py [hunk 3]
```diff
@@ -110,7 +110,7 @@ def _receive(self):
         elif isinstance(msg, ua.Acknowledge):
             self._call_callback(0, msg)
         elif isinstance(msg, ua.ErrorMessage):
-            self.logger.warning("Received an error: %s", msg)
+            self._call_callback(0, ua.UaStatusCodeError(msg.Error.value))
```
Comment (oroulet) [inline opcua/client/ua_client.py]: this might also be correct. but I am a bit worried of the consequences since it might change a lot of things

File: opcua/client/client.py (hunk 1 above)
Comment (urianchang) [reply opcua/client/client.py]: Good point. I've removed the unnecessary try-except and now see the socket being closed in the logs. Thank you!

```
cryptography is not installed, use of crypto disabled
cryptography is not installed, use of crypto disabled
cryptography is not installed, use of crypto disabled
2018-12-05 14:17:21.634 [opcua.client.ua_client.Socket] 140735838761856 INFO - opening connection
2018-12-05 14:17:21.637 [opcua.client.ua_client.Socket] 123145482985472 INFO - Thread started
2018-12-05 14:17:21.638 [opcua.uaprotocol] 123145482985472 DEBUG - Waiting for header
2018-12-05 14:17:21.646 [opcua.uaprotocol] 123145482985472 INFO - received header: Header(type:b'ERR', chunk_type:b'F', body_size:123, channel:0)
2018-12-05 14:17:21.646 [opcua.uaprotocol] 123145482985472 DEBUG - Waiting for header
2018-12-05 14:17:21.646 [opcua.client.ua_client.Socket] 140735838761856 INFO - Request to close socket received
2018-12-05 14:17:21.647 [opcua.client.ua_client.Socket] 123145482985472 INFO - Socket has closed connection
2018-12-05 14:17:21.647 [opcua.client.ua_client.Socket] 123145482985472 INFO - Thread ended
2018-12-05 14:17:21.647 [opcua.client.ua_client.Socket] 140735838761856 INFO - Socket closed, waiting for receiver thread to terminate...
2018-12-05 14:17:21.647 [opcua.client.ua_client.Socket] 140735838761856 INFO - Done closing socket: Receiving thread terminated, socket disconnected
2018-12-05 14:17:21.647 [__main__] 140735838761856 INFO - OPC UA server responded with error BadTcpEndpointUrlInvalid: The server does not recognize the QueryString specified.
```

File: opcua/client/client.py (hunk 2 above)
Comment (urianchang) [reply opcua/client/client.py]: A successful `send_hello` returns an [Acknowledge Message](https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/ua/uaprotocol_hand.py#L81-L97). Looking at Figure 13 and 14 in OPC UA Part 6, 7.1.3, when the client receives the acknowledge message, it'll send a `Open Secure Channel Request`. Not sure if this is helpful, but in the logs, I see this as the header for the `ACK` message:

```
2018-12-05 13:50:31.292 [opcua.uaprotocol] 123145336487936 INFO - received header: Header(type:b'ACK', chunk_type:b'F', body_size:20, channel:0)
```

On a side note, is it fair to assume that a [`UaError`](https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/ua/uaerrors/_base.py#L23-L24) is a client-side error?

File: opcua/client/ua_client.py (hunk 3 above)
Comment (urianchang) [reply opcua/client/ua_client.py]: Yea, I'm not too sure how to handle the `request_id` parameter for the [`_call_callback()` method](https://github.com/FreeOpcUa/python-opcua/blob/master/opcua/client/ua_client.py#L61-L66) because it seemed like the `request_id` was created by the client for tracking message parsing. Reading the "Error handling" section of the OPC UA docs (Part 6, 7.1.5), an *Error Message* should close the *TransportConnection* (which sends the *Hello/Acknowledge Messages*). Since the *Acknowledge* message uses `0`, I thought it'd be safe to use `0` because if the hello returns an error, the error would take the place of the acknowledgement; and if some later request returned an error, setting `0` with the error could prompt the client to close the *TransportConnection* gracefully. For example, if the client gets an error while sending messages, it can save the `_callbackmap` to a temporary file for debugging. I think right now the client reacts to when the Server closes the connection and disconnects the socket, clearing the `_callbackmap`.

**Comment by urianchang** [comment]
FYI: The Scrutinizer [test failure](https://scrutinizer-ci.com/g/FreeOpcUa/python-opcua/inspections/d6558ae1-4540-4212-877f-4c153a02763f) appears with some other PR's.

File: opcua/client/client.py (hunk 2 above)
Comment (oroulet) [reply opcua/client/client.py]: UaError in an exception created from an OPC UA Error, you can raise that Error on client or server

File: opcua/client/ua_client.py (hunk 3 above)
Comment (oroulet) [reply opcua/client/ua_client.py]: I think you should re-enable the logging waring and in fact call logger.fatal as receiving such a message is fatal, as I understand from doc. In fact maybe we should even close the connection ourself at that point, not sure...

File: opcua/common/connection.py [hunk 4]
```diff
@@ -290,7 +290,6 @@ def receive_from_header_and_body(self, header, body):
             return msg
         elif header.MessageType == ua.MessageType.Error:
             msg = struct_from_binary(ua.ErrorMessage, body)
-            logger.warning("Received an error: %s", msg)
```
Comment (oroulet) [inline opcua/common/connection.py]: why did you remove logging here? I know these libraries logs too much but errors should happen that often...

File: opcua/client/ua_client.py (hunk 3 above)
Comment (oroulet) [reply opcua/client/ua_client.py]: @urianchang can you re-enable the warnings? or did you have a very good reason to remove them? see comments above

File: opcua/common/connection.py (hunk 4 above)
Comment (urianchang) [reply opcua/common/connection.py]: I thought it'd be a duplicate, but I've restored it.

File: opcua/client/ua_client.py (hunk 3 above)
Comment (urianchang) [reply opcua/client/ua_client.py]: I've re-enabled the warnings (and elevated this one to fatal).

Comment (urianchang) [reply opcua/client/ua_client.py]: @oroulet, sorry about the delayed response!

Comment (oroulet) [reply opcua/client/ua_client.py]: @urianchang I was even more late... I forgot about that one...
